### ISSUE TITLE
Investigate what is required to support cgroups v2 in libnvidia-container

### ISSUE THREAD

**Issue by arnaldo2792** [bottlerocket-os/bottlerocket#2504]
`libnvidia-container` added support for cgroups v2 in versions >= [1.8.0]. In the 1.10.0 Bottlerocket release, we migrated to `libnvidia-container` 1.11.0 but we didn't compile the library with support for cgroups v2. We should investigate what it takes to enable cgroups v2 support in `libnvidia-container`.

[1.8.0]: https://github.com/NVIDIA/libnvidia-container/releases/tag/v1.8.0

**Comment by arnaldo2792** [comment]
I want to clarify a few things based on the research I've done.

In the 1.8.0 released linked above, the following projects were modified to support cgroups v2:

- `nvidia-container-toolkit`
- `nvidia-container-runtime`
- `libnvidia-container`
- `nvidia-docker`

In Bottlerocket, only `libnvidia-container` and `nvidia-container-toolkit` are used. The other two projects aren't required since we have our own "runtime" ([shimpei] + [oci-add-hooks]) that modifies the container's OCI spec to inject the prestart hook used to mount devices, libraries and binaries, along with setting up the cgroups. Thus, any changes to said projects don't affect how Bottlerocket configures NVIDIA GPUs for containers.

Nonetheless, I was concerned about a [new feature] in `nvidia-container-runtime` that allows a systems administrator specify what files from the host's root filesystem should be mounted into the container using a CSV file. We don't have to support something like this immediately in Bottlerocket since this new feature was added for a specific platform:

> This mode is primarily targeted at Tegra-based systems without NVML available.

From the two projects we do care, only `libnvidia-container` was modified to support cgroups v2. A [new go library] was created which is in charge of setting up the cgroups for the container. I haven't figure out if this library supports cgroups v1 as well, if not we will have to change the RPM spec file to conditionally compile the library depending on the variant being built.

[shimpei]: https://github.com/bottlerocket-os/bottlerocket/tree/develop/sources/shimpei
[oci-add-hooks]: https://github.com/awslabs/oci-add-hooks
[new feature]: https://github.com/NVIDIA/nvidia-container-toolkit/tree/main/cmd/nvidia-container-runtime#runtime-mode
[new go library]: https://github.com/NVIDIA/libnvidia-container/blob/90bc7a540add5857fcd30c1f4f3bede9b5b06f43/src/nvcgo/main.go

**Comment by markusboehme** [comment]
I have `libnvidia-container` building with cgroup v2 support now. The current state of the work is available in [the `feature/nvidia-on-cgroup2` branch of my fork](https://github.com/markusboehme/bottlerocket/tree/feature/nvidia-on-cgroup2).

We do need to make use of the new Go library (`nvcgo`/`libnvidia-container-go`). This is a compile-time option. Since the library supports both cgroup v1 and v2, there's no need for conditional building.

While the above state of the branch works fine for a cgroup v1 host, on a cgroup v2 host I currently get the following error when trying to create a pod with a GPU assigned:

```
nvidia-container-cli: mount error: failed to add device rules: unable to generate new device filter program from existing programs: unable to create new device filters program: load program: invalid argument: last insn is not an exit or jmp
processed 0 insns (limit 1000000) max_states_per_insn 0 total_states 0 peak_states 0 mark_read 0: unknown
```

The kernel's eBPF verifier is rejecting the program that allows/disallows GPU access from the container because it might run off into the weeds. The relevant code is in [`cgroup/ebpf.go` of `libnvidia-container`](https://github.com/NVIDIA/libnvidia-container/blob/main/src/nvcgo/internal/cgroup/ebpf.go) (`PrependDeviceFilter`), called by `AddDeviceRules` from [`cgroup/v2.go` of `libnvidia-container`](https://github.com/NVIDIA/libnvidia-container/blob/main/src/nvcgo/internal/cgroup/v2.go).

On first skimming, the code takes care to terminate programs with an explicit return. This will probably require some instrumentation to see the assembled program sent to the kernel.

Worth noting, the problem has been [reported upstream](https://github.com/NVIDIA/libnvidia-container/issues/176) already. Alas, there's been no conclusion yet.

**Comment by markusboehme** [comment]
tl;dr: `echo 1 >/sys/proc/net/core/bpf_jit_harden` to fix by disabling hardening for eBPF programs loaded by privileged users. 

Bottlerocket enables eBPF JIT hardening for both privileged and unprivileged users by default. One of the hardening measures is a constant blinding pass over eBPF bytecode loaded into the kernel, which applies slight modifications to the bytecode that preserve semantics but decrease attacker control over possible instruction sequences ending up in executable kernel memory. As a side effect, programs that have been blinded cannot be dumped to user space again.

It is the inability to dump eBPF programs to user space that is causing problems for `libnvidia-container-go`. When allowing GPU access to a container, it prepends new filters to the existing program (I assume put in place for the cgroup by runc). If constant blinding has been applied to the program, `libnvidia-container-go` will prepend the new filter to [a buffer of zeros](https://github.com/NVIDIA/libnvidia-container/blob/7aa2c9b2d0fe5450c0a0dfc903899e1fa547fc63/src/nvcgo/vendor/github.com/cilium/ebpf/info.go#L136). "All zeros" happens to be a valid eBPF instruction encoding (loading a constant 32-bit 0 into register 0), so when the modified program is loaded back into the kernel, the eBPF verifier will only notice there's code path that does not explicitly terminate. This results in the error seen above, with the new device filter program being rejected on the grounds of "last insn is not an exit or jmp".

I'll bring what I found to the upstream issue. We may work around it by setting the `net.core.bpf_jit_harden` sysctl to `1`, only applying the hardening measures for eBPF programs to those loaded by unprivileged users (note that we already disable eBPF use by unprivileged users, so the effect would be similar as to setting `net.core.bpf_jit_harden` to `0`). The change could be scoped to NVIDIA AMIs only, similar to how we already disable kernel lockdown mode for these.
